src/doom/p_enemy.c

```diff
@@ -235,34 +235,26 @@ boolean P_CheckMissileRange (mobj_t* actor)
 
     dist >>= FRACBITS;
 
-    if (actor->type == MT_VILE)
-    {
-	if (dist > 14*64)	
-	    return false;	// too far away
-    }
-	
-
-    if (actor->type == MT_UNDEAD)
-    {
-	if (dist < 196)	
-	    return false;	// close for fist attack
-	dist >>= 1;
-    }
+    // generalization of the Arch Vile's different attack range
+    if (actor->info->maxattackrange > 0 && dist > actor->info->maxattackrange)
+	return false;	// too far away
 	
+    // generalization of the Revenant's different melee threshold
+    if (actor->info->meleethreshold > 0 && dist < actor->info->meleethreshold)
+	return false;	// close for fist attack
 
     if (actor->type == MT_CYBORG
 	|| actor->type == MT_SPIDER
+	|| actor->type == MT_UNDEAD
```

Comment:
Not very convincing to "un-hard-code" some parts of this while at the same time hard-coding some others. But I guess this is already discussed in the forum thread? The `MT_UNDEAD` case could already be covered by the `(actor->info->meleethreshold > 0)` condition as in the original logic, right?

Reply:
Doing it that way would mean that every enemy with a melee threshold would double its chance of firing missiles.

The original logic was adding 2 different special behaviors to the revenant, the change was intended to de-hardcode one of them (the distance the revenant uses to decide whether to punch or to fire). Without touching yet its double firing chance (which in ZDoom/Eternity is coded into a flag).

I'm hoping for more feedback in the forum thread on whether adding a "Missile chance multiplier" makes sense and in what form (the bit shift here is presumably more efficient than a multiplication, which is why replacing it might be a bit touchy).

Reply:
The problem is that the flags bit-field is already pretty constraint and adding another flag there is likely to cause incompatibilities across ports - but you already know that. I hope that consensus can be reached that this requires another explicit thing property that can be changed by dehacked.

Reply:
> The original logic was adding 2 different special behaviors to the revenant, the change was intended to de-hardcode one of them (the distance the revenant uses to decide whether to punch or to fire). Without touching yet its double firing chance (which in ZDoom/Eternity is coded into a flag).

One could also read the Vanilla logic here that all monsters with a melee threshold also have their firing chance modified - with "all monsters" being only the Revenant if no dehackery is applied.

Reply:
That's just a different way of hard-coding the behavior.
The melee threshold should be independent of the firing multiplier, imho.

Reply:
You are probably right. 👍